File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        set<vector<int>> ans;
        unordered_set<int> dups;
        unordered_map<int, int> seen;
        
        for (int i = 0; i < nums.size(); ++i){
            if (!dups.count(nums[i])){
                dups.insert(nums[i]);
                for (int j = i + 1; j < nums.size(); ++j){
                    int complement = -(nums[i] + nums[j]);
                    if (seen.count(complement) && seen[complement] == i){
                        vector<int> triplet = {nums[i], nums[j], complement};
                        sort(triplet.begin(), triplet.end());
                        ans.insert(triplet);
                    }
                    seen[nums[j]] = i;
                }
            }
            
        }
        
        return vector<vector<int>>(ans.begin(), ans.end());
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans, dups = set(), set()
        seen = dict()
        
        for i, val1 in enumerate(nums):
            if val1 not in dups:
                dups.add(val1)
                for j, val2 in enumerate(nums[i + 1:]):
                    complement = - (val1 + val2)
                    if complement in seen and seen[complement] == i:
                        ans.add(tuple(sorted((val1, val2, complement))))
                    seen[val2] = i
        
        return ans